### DISCUSSION TITLE
How to generate random subtype instances of a given type?

### DISCUSSION THREAD

**Discussion by bash-spbu** [instancio/instancio#1126] [Q&A] [answered]
Hello!

Could you please suggest how can I get an instance of some random subtype among specified ones for a given super type?
Also I would want to get list of such random subtype instances of the same supertype.

Something like this:

```java
interface Vehicle {}

class Car     implements Vehicle {}
class Bicycle implements Vehicle {}
class Boat    implements Vehicle {}

// doesn't work :c
final var carOrBicycleOrBoat = Instancio.of(Vehicle.class)
  .subtype(all(Vehicle.class), Car.class)
  .subtype(all(Vehicle.class), Bicycle.class)
  .subtype(all(Vehicle.class), Boat.class)
  .create();

final var listOfRandomVehicles = ???;
```

**Comment by armandino** [comment, marked as answer]
@bash-spbu Instancio selects random subtypes only for `sealed` abstract types. You can find more info here:

https://www.instancio.org/user-guide/#subtype-mapping

Since you're not using `sealed` types, you can achieve something similar by creating a helper method:

```java
static Supplier<Vehicle> vehicleSupplier() {
    return () -> {
        Class<?> subtype = Instancio.gen()
                .oneOf(Car.class, Bicycle.class, Boat.class)
                .get();

        return Instancio.of(Vehicle.class)
                .subtype(all(Vehicle.class), subtype)
                .create();
    };
}

List<Vehicle> vehicles = Stream.generate(vehicleSupplier())
        .limit(100)
        .collect(Collectors.toList());

assertThat(vehicles).hasOnlyElementsOfTypes(Car.class, Bicycle.class, Boat.class);
```

>  Also I would want to get list of such random subtype instances of the same supertype.

For this, you can implement the SPI. Please check out the `type-resolver-sample` in https://github.com/instancio/instancio-samples for an example.

**Reply by bash-spbu** [reply]
Thank you very much for the detailed answer!
